autopush/endpoint.py

```diff
@@ -110,29 +113,38 @@ def _process_uaid(self, result):
 
         if node_id:
             # Attempt a delivery if they are connected
-            payload = json.dumps([{"channelID": self.chid,
-                                   "version": self.version,
-                                   "data": self.data}])
-            d = deferToThread(
-                self.ap_settings.requests.put,
-                node_id + "/push/" + self.uaid,
-                data=payload
-            )
+            d = self._send_notification(node_id)
             d.addCallback(self._process_routing, result)
             d.addErrback(self._error_response)
         else:
             self._save_notification()
 
-    def _process_routing(self, result, item):
+    def _send_notification(self, node_id):
+        payload = json.dumps([{"channelID": self.chid,
+                               "version": self.version,
+                               "data": self.data}])
+        return self.ap_settings.agent.request(
+            "PUT",
+            node_id + "/push/" + self.uaid,
+            bodyProducer=FileBodyProducer(StringIO(payload)),
+        ).addCallback(IgnoreBody.ignore)
```

Comment:
Not quite...if there's a request error, `IgnoreBody.ignore` won't be called and the error will be forwarded to the next registered errback (`self._error_response`), just like JavaScript promises. I can add a test.